cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,483 @@
+				t.Errorf("got %q, want %q", rules[0].PortRange, inboundRules[0].PortRange)
+			}
+			if !cmp.Equal(rules[0].Protocol, inboundRules[0].Protocol) {
+				t.Errorf("got %q, want %q", rules[0].Protocol, inboundRules[0].Protocol)
+			}
+		})
+	}
+}
+
+func TestFirewallController_run(t *testing.T) {
+	testcases := []struct {
+		name                           string
+		fwCache                        firewallCache
+		expectedGodoFirewallCreateResp func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallGetResp    func(context.Context, string) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallListResp   func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallUpdateResp func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error)
+		fwReconcileFreq                time.Duration
+		expectedError                  error
+	}{
+		{
+			name:    "check for data race with high frequency",
+			fwCache: newFakeFirewallCache(fakeInboundRule),
+			expectedGodoFirewallCreateResp: func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeOKResponse(), nil
+			},
+			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(fakeWorkerFWName, fakeInboundRule), newFakeOKResponse(), nil
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return nil, newFakeOKResponse(), nil
+			},
+			expectedGodoFirewallUpdateResp: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(fakeWorkerFWName, fakeInboundRule), newFakeOKResponse(), nil
+			},
+			// force a high frequency to increase chance of discovering data races
+			fwReconcileFreq: time.Duration(0),
+		},
+	}
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			// setup
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			defer cancel()
+			var wg sync.WaitGroup
+
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					listFunc:   test.expectedGodoFirewallListResp,
+					updateFunc: test.expectedGodoFirewallUpdateResp,
+					createFunc: test.expectedGodoFirewallCreateResp,
+					getFunc:    test.expectedGodoFirewallGetResp,
+				},
+			)
+			fwManagerOp := newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for ctx.Err() == nil { // context has not been terminated
+					fc.ensureReconciledFirewall(ctx)
+				}
+			}()
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				fc.Run(ctx.Done(), fwManagerOp, test.fwReconcileFreq)
+			}()
+
+			wg.Wait()
```

Comment:
Can we add a comment after this line explaining that we do not assert on anything because the goal is to catch data races (which would let the test fail entirely through the data race detector)?